assets/app/view/game/part/town_dot.rb

```diff
@@ -69,32 +69,54 @@ def preferred_render_locations
               },
             ]
           else
-            @tile.towns.size > 1 ? OFFSET_TOWNS : CENTER_TOWN
+            # see if any other towns are also in the center
+            @tile.towns.count { |t| !@tile.preferred_city_town_edges[t] } > 1 ? OFFSET_TOWNS : CENTER_TOWN
           end
         end
 
         def render_revenue
-          revenues = @tile.towns.first.revenue.values.uniq
-          return unless revenues.one? && @town.paths.any?
+          revenues = @town.uniq_revenues
+          return if revenues.size > 1
 
           revenue = revenues.first
+          return if !@town.halt? && revenue.zero?
+
+          x = render_location[:x]
+          y = render_location[:y]
 
           angle = layout == :pointy ? -60 : 0
 
           displacement = @edge ? REVENUE_EDGE_DISPLACEMENT : REVENUE_DISPLACEMENT
 
           increment_weight_for_regions(REVENUE_REGIONS[layout])
-          h(:g, { attrs: { transform: "rotate(#{angle})" } }, [
-              h(:g, { attrs: { transform: "translate(#{displacement} 0) rotate(#{-angle})" } }, [
-                  h(Part::SingleRevenue,
-                    revenue: revenue,
-                    transform: rotation_for_layout),
+          if @town.halt?
```

Comment:
seems like there's a lot of duplicate code here that could be combined

Reply:
Ok